File: electron_node/docs/AGGREGATOR/aggregator_decision.py
"""
Aggregator core decision logic: Text Incompleteness Score + Language Stability Gate
Copy-paste friendly. No third-party deps.
"""

from __future__ import annotations
from dataclasses import dataclass
from typing import Optional, Literal

Mode = Literal["offline", "room"]
StreamAction = Literal["MERGE", "NEW_STREAM"]


@dataclass
class LangProbs:
    top1: str
    p1: float
    top2: Optional[str] = None
    p2: Optional[float] = None


@dataclass
class UtteranceInfo:
    text: str
    start_ms: int
    end_ms: int
    lang: LangProbs
    quality_score: Optional[float] = None
    is_final: bool = False
    is_manual_cut: bool = False


@dataclass
class AggregatorTuning:
    strong_merge_ms: int
    soft_gap_ms: int
    hard_gap_ms: int

    lang_stable_p: float
    lang_switch_margin: float
    lang_switch_requires_gap_ms: int

    score_threshold: int
    w_short: int
    w_very_short: int
    w_gap_short: int
    w_no_strong_punct: int
    w_ends_with_connective: int
    w_low_quality: int
    low_quality_threshold: float

    short_cjk_chars: int
    very_short_cjk_chars: int
    short_en_words: int
    very_short_en_words: int

    commit_interval_ms: int
    commit_len_cjk: int
    commit_len_en_words: int


def default_tuning(mode: Mode) -> AggregatorTuning:
    is_room = mode == "room"
    return AggregatorTuning(
        strong_merge_ms=600 if is_room else 700,
        soft_gap_ms=1000 if is_room else 1500,
        hard_gap_ms=1500 if is_room else 2000,

        lang_stable_p=0.80,
        lang_switch_margin=0.18 if is_room else 0.15,
        lang_switch_requires_gap_ms=500 if is_room else 600,

        score_threshold=3,
        w_short=2,
        w_very_short=3,
        w_gap_short=2,
        w_no_strong_punct=1,
        w_ends_with_connective=1,
        w_low_quality=1,
        low_quality_threshold=0.50 if is_room else 0.45,

        short_cjk_chars=9 if is_room else 10,
        very_short_cjk_chars=4,
        short_en_words=5 if is_room else 6,
        very_short_en_words=3,

        commit_interval_ms=900 if is_room else 1400,
        commit_len_cjk=22 if is_room else 30,
        commit_len_en_words=10 if is_room else 12,
    )


def decide_stream_action(
    prev: Optional[UtteranceInfo],
    curr: UtteranceInfo,
    mode: Mode,
    tuning: Optional[AggregatorTuning] = None,
) -> StreamAction:
    tuning = tuning or default_tuning(mode)
    if prev is None:
        return "NEW_STREAM"

    gap_ms = max(0, curr.start_ms - prev.end_ms)

    # Hard rules
    if curr.is_final or curr.is_manual_cut:
        return "NEW_STREAM"
    if gap_ms >= tuning.hard_gap_ms:
        return "NEW_STREAM"

    # Language stability gate
    if is_lang_switch_confident(prev.lang, curr.lang, gap_ms, tuning):
        return "NEW_STREAM"

    # Strong merge if extremely continuous
    if gap_ms <= tuning.strong_merge_ms:
        return "MERGE"

    score = text_incompleteness_score(prev, curr, gap_ms, tuning)
    if score >= tuning.score_threshold and gap_ms <= tuning.soft_gap_ms:
        return "MERGE"
    return "NEW_STREAM"


def is_lang_switch_confident(
    prev_lang: LangProbs,
    curr_lang: LangProbs,
    gap_ms: int,
    tuning: AggregatorTuning,
) -> bool:
    if gap_ms <= tuning.lang_switch_requires_gap_ms:
        return False
    if prev_lang.p1 < tuning.lang_stable_p or curr_lang.p1 < tuning.lang_stable_p:
        return False
    if prev_lang.top1 == curr_lang.top1:
        return False
    p2 = curr_lang.p2 or 0.0
    return (curr_lang.p1 - p2) >= tuning.lang_switch_margin


def text_incompleteness_score(
    prev: UtteranceInfo,
    curr: UtteranceInfo,
    gap_ms: int,
    tuning: AggregatorTuning,
) -> int:
    score = 0

    is_cjk = looks_like_cjk(curr.text)
    cjk_chars = count_cjk_chars(curr.text) if is_cjk else 0
    en_words = count_words(curr.text) if not is_cjk else 0

    short = (cjk_chars < tuning.short_cjk_chars) if is_cjk else (en_words < tuning.short_en_words)
    very_short = (cjk_chars < tuning.very_short_cjk_chars) if is_cjk else (en_words < tuning.very_short_en_words)

    if very_short:
        score += tuning.w_very_short
    elif short:
        score += tuning.w_short

    if gap_ms < (tuning.strong_merge_ms + 200):
        score += tuning.w_gap_short

    if not ends_with_strong_sentence_punct(curr.text):
        score += tuning.w_no_strong_punct

    if ends_with_connective_or_filler(curr.text):
        score += tuning.w_ends_with_connective

    q = curr.quality_score if curr.quality_score is not None else 1.0
    if q < tuning.low_quality_threshold:
        score += tuning.w_low_quality

    if (not ends_with_strong_sentence_punct(prev.text)) and gap_ms <= tuning.soft_gap_ms:
        score += 1

    return score


def should_commit(
    pending_text: str,
    last_commit_ts_ms: int,
    now_ms: int,
    mode: Mode,
    tuning: Optional[AggregatorTuning] = None,
) -> bool:
    tuning = tuning or default_tuning(mode)
    if (now_ms - last_commit_ts_ms) >= tuning.commit_interval_ms:
        return True
    if looks_like_cjk(pending_text):
        return count_cjk_chars(pending_text) >= tuning.commit_len_cjk
    return count_words(pending_text) >= tuning.commit_len_en_words


# ---------- helpers ----------

def ends_with_strong_sentence_punct(s: str) -> bool:
    t = s.strip()
    return bool(t) and t[-1] in "。！？.!?；;"


def looks_like_cjk(s: str) -> bool:
    for ch in s:
        u = ord(ch)
        if (0x3040 <= u <= 0x30FF) or (0x3400 <= u <= 0x4DBF) or (0x4E00 <= u <= 0x9FFF) or (0xAC00 <= u <= 0xD7AF):
            return True
    return False


def count_cjk_chars(s: str) -> int:
    n = 0
    for ch in s:
        u = ord(ch)
        if (0x3040 <= u <= 0x30FF) or (0x3400 <= u <= 0x4DBF) or (0x4E00 <= u <= 0x9FFF) or (0xAC00 <= u <= 0xD7AF):
            n += 1
    return n


def count_words(s: str) -> int:
    return len([w for w in s.strip().split() if w])


def ends_with_connective_or_filler(s: str) -> bool:
    t = s.strip().lower()
    if not t:
        return False

    en = ["and", "but", "so", "because", "then"]
    for w in en:
        if t == w or t.endswith(" " + w):
            return True

    zh = ["然后", "所以", "但是", "就是", "那个", "嗯", "呃"]
    ja = ["で", "から", "けど", "えっと"]
    ko = ["그리고", "근데", "그래서", "어", "음"]
    for w in zh + ja + ko:
        if t.endswith(w):
            return True
    return False

File: electron_node/electron-node/main/src/aggregator/aggregator-decision.ts
/* Aggregator core decision logic: Text Incompleteness Score + Language Stability Gate
   Copy-paste friendly. No external deps.
*/

export type Mode = "offline" | "room" | "two_way";
export type StreamAction = "MERGE" | "NEW_STREAM";

export interface LangProbs {
  top1: string;
  p1: number;
  top2?: string;
  p2?: number;
}

export interface UtteranceInfo {
  text: string;
  startMs: number;
  endMs: number;
  lang: LangProbs;
  qualityScore?: number;
  isFinal?: boolean;
  isManualCut?: boolean;
  isPauseTriggered?: boolean;
  isTimeoutTriggered?: boolean;
}

export interface AggregatorTuning {
  strongMergeMs: number;
  softGapMs: number;
  hardGapMs: number;

  langStableP: number;
  langSwitchMargin: number;
  langSwitchRequiresGapMs: number;

  scoreThreshold: number;
  wShort: number;
  wVeryShort: number;
  wGapShort: number;
  wNoStrongPunct: number;
  wEndsWithConnective: number;
  wLowQuality: number;
  lowQualityThreshold: number;

  shortCjkChars: number;
  veryShortCjkChars: number;
  shortEnWords: number;
  veryShortEnWords: number;

  commitIntervalMs: number;
  commitLenCjk: number;
  commitLenEnWords: number;
}

export function defaultTuning(mode: Mode): AggregatorTuning {
  const isRoom = mode === "room";
  const isTwoWay = mode === "two_way";
  // 双向互译模式使用与 room 模式相同的参数
  const useRoomParams = isRoom || isTwoWay;
  return {
    // 质量优化：提高 strongMergeMs，降低 softGapMs，降低 scoreThreshold
    // 让更多片段被 MERGE，减少独立翻译，提升翻译质量
    strongMergeMs: useRoomParams ? 800 : 1000,  // 提高：让更多短片段被 MERGE
    softGapMs: useRoomParams ? 1000 : 1200,     // 降低：让更多片段被 MERGE
    hardGapMs: useRoomParams ? 1500 : 2000,

    langStableP: 0.8,
    langSwitchMargin: useRoomParams ? 0.18 : 0.15,
    langSwitchRequiresGapMs: useRoomParams ? 500 : 600,

    scoreThreshold: 2.5,  // 降低：让更多片段被 MERGE
    wShort: 2,
    wVeryShort: 3,
    wGapShort: 2,
    wNoStrongPunct: 1,
    wEndsWithConnective: 1,
    wLowQuality: 1,
    lowQualityThreshold: useRoomParams ? 0.5 : 0.45,

    shortCjkChars: useRoomParams ? 9 : 10,
    veryShortCjkChars: 4,
    shortEnWords: useRoomParams ? 5 : 6,
    veryShortEnWords: 3,

    // 优化：平衡延迟和质量
    // 问题：当前参数导致句子中间截断
    // 调整：提高 commit_interval_ms 和 commit_len，减少句子中间截断
    // 原值：offline 800ms/25字/10词，room 600ms/20字/8词
    // 新值：offline 1200ms/30字/12词，room/two_way 900ms/25字/10词
    commitIntervalMs: useRoomParams ? 900 : 1200,  // 提高：减少句子中间截断
    commitLenCjk: useRoomParams ? 25 : 30,         // 提高：减少短句被提前提交
    commitLenEnWords: useRoomParams ? 10 : 12,     // 提高：减少短句被提前提交
  };
}

export function decideStreamAction(
  prev: UtteranceInfo | null,
  curr: UtteranceInfo,
  mode: Mode,
  tuning: AggregatorTuning = defaultTuning(mode)
): StreamAction {
  if (!prev) return "NEW_STREAM";

  const gapMs = Math.max(0, curr.startMs - prev.endMs);

  // Hard rules
  // P0 修复：只对 manualCut 强制 NEW_STREAM，isFinal 允许 MERGE
  // 因为 P0 只处理 final 结果，如果 isFinal 也强制 NEW_STREAM，会导致无法 MERGE
  if (curr.isManualCut) return "NEW_STREAM";
  if (gapMs >= tuning.hardGapMs) return "NEW_STREAM";

  // Language stability gate
  if (isLangSwitchConfident(prev.lang, curr.lang, gapMs, tuning)) return "NEW_STREAM";

  // Strong merge
  if (gapMs <= tuning.strongMergeMs) return "MERGE";

  const score = textIncompletenessScore(prev, curr, gapMs, tuning);
  if (score >= tuning.scoreThreshold && gapMs <= tuning.softGapMs) return "MERGE";

  return "NEW_STREAM";
}

export function isLangSwitchConfident(
  prevLang: LangProbs,
  currLang: LangProbs,
  gapMs: number,
  tuning: AggregatorTuning
): boolean {
  if (gapMs <= tuning.langSwitchRequiresGapMs) return false;
  if (prevLang.p1 < tuning.langStableP || currLang.p1 < tuning.langStableP) return false;
  if (prevLang.top1 === currLang.top1) return false;

  const p2 = currLang.p2 ?? 0;
  return (currLang.p1 - p2) >= tuning.langSwitchMargin;
}

export function textIncompletenessScore(
  prev: UtteranceInfo,
  curr: UtteranceInfo,
  gapMs: number,
  tuning: AggregatorTuning
): number {
  let score = 0;

  const isCjk = looksLikeCjk(curr.text);
  const cjkChars = isCjk ? countCjkChars(curr.text) : 0;
  const enWords = !isCjk ? countWords(curr.text) : 0;

  const short = isCjk ? cjkChars < tuning.shortCjkChars : enWords < tuning.shortEnWords;
  const veryShort = isCjk ? cjkChars < tuning.veryShortCjkChars : enWords < tuning.veryShortEnWords;

  if (veryShort) score += tuning.wVeryShort;
  else if (short) score += tuning.wShort;

  if (gapMs < (tuning.strongMergeMs + 200)) score += tuning.wGapShort;

  if (!endsWithStrongSentencePunct(curr.text)) score += tuning.wNoStrongPunct;

  if (endsWithConnectiveOrFiller(curr.text)) score += tuning.wEndsWithConnective;

  const q = curr.qualityScore ?? 1.0;
  if (q < tuning.lowQualityThreshold) score += tuning.wLowQuality;

  if (!endsWithStrongSentencePunct(prev.text) && gapMs <= tuning.softGapMs) score += 1;

  return score;
}

export function shouldCommit(
  pendingText: string,
  lastCommitTsMs: number,
  nowMs: number,
  mode: Mode,
  tuning: AggregatorTuning = defaultTuning(mode)
): boolean {
  const elapsed = nowMs - lastCommitTsMs;
  if (elapsed >= tuning.commitIntervalMs) return true;

  const isCjk = looksLikeCjk(pendingText);
  if (isCjk) return countCjkChars(pendingText) >= tuning.commitLenCjk;
  return countWords(pendingText) >= tuning.commitLenEnWords;
}

/* Helpers */

export function endsWithStrongSentencePunct(s: string): boolean {
  const t = s.trim();
  if (!t) return false;
  return /[。！？.!?；;]$/.test(t);
}

export function looksLikeCjk(s: string): boolean {
  return /[\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uac00-\ud7af]/.test(s);
}

export function countCjkChars(s: string): number {
  const m = s.match(/[\u3040-\u30ff\u3400-\u4dbf\u4e00-\u9fff\uac00-\ud7af]/g);
  return m ? m.length : 0;
}

export function countWords(s: string): number {
  const t = s.trim();
  if (!t) return 0;
  return t.split(/\s+/).filter(Boolean).length;
}

export function endsWithConnectiveOrFiller(s: string): boolean {
  const t = s.trim().toLowerCase();
  if (!t) return false;

  const en = ["and", "but", "so", "because", "then"];
  const zh = ["然后", "所以", "但是", "就是", "那个", "嗯", "呃"];
  const ja = ["で", "から", "けど", "えっと"];
  const ko = ["그리고", "근데", "그래서", "어", "음"];

  for (const w of en) if (t.endsWith(" " + w) || t === w) return true;
  for (const w of zh) if (t.endsWith(w)) return true;
  for (const w of ja) if (t.endsWith(w)) return true;
  for (const w of ko) if (t.endsWith(w)) return true;

  return false;
}

